### ISSUE TITLE
Nerf trajopt example on cuda

### ISSUE THREAD

**Issue by adamdai** [Tim-Salzmann/l4casadi#34]
Hey Tim,

I'm trying to get the nerf trajectory optimization example running on cuda. I was able to get it running on cpu, and edited lines 171-179 as follows: 
```
# --------------------------------- Load NERF -------------------------------- #
    model = DensityNeRF()
    model_path = os.path.join(os.path.dirname(__file__), "nerf_model.tar")
    model.load_state_dict(
        torch.load(model_path, map_location="cuda")["network_fn_state_dict"],
        strict=False,
    )
    # -------------------------- Create L4CasADi Module -------------------------- #
    l4c_nerf = l4c.L4CasADi(model, device="cuda")
```

However, when I run this, the second nlp solver seems to hang:
```
(l4casadi) adam@adam-ubuntu:~/Sandbox/l4casadi/examples/nerf_trajectory_optimization$ python nerf_trajectory_optimization.py 
This is Ipopt version 3.14.11, running with linear solver MUMPS 5.4.1.

Number of nonzeros in equality constraint Jacobian...:       36
Number of nonzeros in inequality constraint Jacobian.:     2700
Number of nonzeros in Lagrangian Hessian.............:       90

Total number of variables............................:       18
                     variables with only lower bounds:        0
                variables with lower and upper bounds:        0
                     variables with only upper bounds:        0
Total number of equality constraints.................:        6
Total number of inequality constraints...............:      300
        inequality constraints with only lower bounds:        0
   inequality constraints with lower and upper bounds:      300
        inequality constraints with only upper bounds:        0

iter    objective    inf_pr   inf_du lg(mu)  ||d||  lg(rg) alpha_du alpha_pr  ls
   0  1.0161843e+02 1.19e+00 1.00e+02  -1.0 0.00e+00    -  0.00e+00 0.00e+00   0
   1  3.5003351e+01 4.43e-08 2.22e+00  -1.0 4.80e+03    -  3.01e-01 1.00e+00f  1
   2  1.0827344e+01 4.02e-08 1.13e+00  -1.0 2.26e+03    -  8.49e-01 1.00e+00f  1
   3  2.7440984e+00 1.55e-09 2.47e-11  -1.0 1.85e+03    -  1.00e+00 1.00e+00f  1
   4  4.0409223e-01 2.04e-08 4.75e-01  -2.5 1.44e+03    -  9.40e-01 1.00e+00f  1
   5  7.3485968e-02 6.77e-10 2.87e-11  -2.5 1.15e+03    -  1.00e+00 1.00e+00f  1
   6  4.1873740e-02 4.36e-10 1.37e-11  -3.8 5.28e+02    -  1.00e+00 1.00e+00f  1
   7  4.0472866e-02 6.24e-11 6.51e-12  -3.8 2.85e+02    -  1.00e+00 1.00e+00f  1
   8  4.0419962e-02 9.64e-12 8.72e-12  -5.7 5.41e+01    -  1.00e+00 1.00e+00f  1
   9  4.0419926e-02 6.06e-13 1.36e-11  -8.6 1.88e+00    -  1.00e+00 1.00e+00h  1

Number of Iterations....: 9

                                   (scaled)                 (unscaled)
Objective...............:   2.9568343777355507e-02    4.0419925943644970e-02
Dual infeasibility......:   1.3623588368538719e-11    1.8623445299792427e-11
Constraint violation....:   6.0595972684041044e-13    6.0595972684041044e-13
Variable bound violation:   0.0000000000000000e+00    0.0000000000000000e+00
Complementarity.........:   9.9614675922336054e-09    1.3617326198583336e-08
Overall NLP error.......:   9.9614675922336054e-09    1.3617326198583336e-08


Number of objective function evaluations             = 10
Number of objective gradient evaluations             = 10
Number of equality constraint evaluations            = 10
Number of inequality constraint evaluations          = 10
Number of equality constraint Jacobian evaluations   = 10
Number of inequality constraint Jacobian evaluations = 10
Number of Lagrangian Hessian evaluations             = 9
Total seconds in IPOPT                               = 0.594

EXIT: Optimal Solution Found.
/home/adam/anaconda3/envs/l4casadi/lib/python3.10/site-packages/torch/jit/_check.py:177: UserWarning: The TorchScript type system doesn't support instance-level annotations on empty non-base types in `__init__`. Instead, either 1) use a type annotation in the class body, or 2) wrap the type in `torch.jit.Attribute`.
  warnings.warn(
CUDA is available! Using GPU cuda.
This is Ipopt version 3.14.11, running with linear solver MUMPS 5.4.1.

Number of nonzeros in equality constraint Jacobian...:       36
Number of nonzeros in inequality constraint Jacobian.:     5400
Number of nonzeros in Lagrangian Hessian.............:      171

```

Any thoughts on what the issue might be? I'm using a conda environment with python 3.10 and torch 2.3.0.

**Comment by Tim-Salzmann** [comment]
Hi Adam,

I looked into this. For unknown reasons, the second c++ jit forward pass freezes. It seems like the DensityNeRF model has something which the c++ jit fusion engine can not handle. As of now, I am unsure what exactly the underlying problem is. [1]

For now, the (dirty) fix is to add `torch.jit.set_fusion_strategy([('STATIC', 0)])` at the top of your python script. This will likely affect the performance of the model; however, I am unsure how much.

I'll potentially have another look at this later this week. Please let me know if this unblocks you for now.

Best
Tim

[1] https://discuss.pytorch.org/t/slow-forward-on-traced-graph-on-cuda-2nd-iteration/118445/6

**Comment by adamdai** [comment]
Thanks Tim! Your tip worked, I was able to run the script. 

I'm also curious in general about the performance difference between gpu and cpu. I noticed that in most of the examples gpu tends to perform similar or worse than cpu. Have you also observed this on your end?

**Comment by adamdai** [comment]
Hi Tim,

Got it, that makes sense. Thanks so much for the detailed answer and example!

I did also notice that you are evaluating the NeRF in a for loop (i.e. batch size of 1 as you mentioned). You mentioned that solvers are not designed to take advantage of batching - is there a reason why it isn't possible to evaluate the NeRF density over the whole trajectory as a single batched input? I ask since we are looking into traj opt in implicit maps, and wondering if batching would be possible/provide a speed up.

Thanks,
Adam

**Comment by Tim-Salzmann** [comment]
Hi Adam,

Again, these are very valid points! My argument in the last post that "solvers" can not handle batches was slightly sloppy. In the NeRF example, CasADi is the limiting factor. My understanding is that CasADi does not have a notion of a "batched dimensions". As such, in the NeRF example, if we pass a batched vector ([150, 3]) to the L4CasADi NeRF function, the forward pass will work (output -> [150, 1]). However, as a Jacobian, CasADi will expect the L4CasADi function to return a vector of size [150, 3, 150, 1] when we only want to calculate a vector of [150, 3, 1]. Similarly, for the Hessian ([150, 3, 150, 1, 150, 3, 150, 1] vs [150, 3, 1, 3, 1]). As you can see, this will produce a massive overhead.


Now, I am not an expert in CasADi, and it would make sense to double-check this or ask in the CasADi Github if there is a way to somehow leverage the batching capabilities of Torch/L4CasADi. If there is a way to cleverly leverage torch's batching capabilities (e.g. maybe via the sparsity propagated in CasADi 🤔), I would be more than happy to adapt L4CasADi to make this work.

Best
Tim

**Comment by Tim-Salzmann** [comment]
https://github.com/casadi/casadi/discussions/3715

**Comment by Tim-Salzmann** [comment]
Hi Adam,

I figured out a way to potentially leverage such parallel calls better: https://github.com/casadi/casadi/discussions/3715

While planning to implement this, it will most likely take some time until I find the time to do this properly (this will be some larger changes to L4CasADi)

If this is okay with you, I will close the issue for now.

Best
Tim

**Comment by adamdai** [comment]
Hey Tim,

Thanks for looking into this! This would definitely be very beneficial for our use cases. Yes you can feel free to close this issue, I'll stay tuned for updates and also keep you posted if I find anything on my end.

Best,
Adam
